syncano/models/manager.py

```diff
@@ -0,0 +1,251 @@
+        return self.request()
+
+    def update_or_create(self, *args, **kwargs):
+        data = deepcopy(kwargs.get('data', {}))
+        try:
+            instance = self.update(*args, **kwargs)
+        except self.model.DoesNotExist:
+            data.update(self.properties)
+            data.update(kwargs)
+            instance = self.create(**data)
+        return instance
+
+    # List actions
+
+    def all(self, *args, **kwargs):
+        self._limit = None
+        return self.list(*args, **kwargs)
+
+    def list(self, *args, **kwargs):
+        self.method = 'GET'
+        self.endpoint = 'list'
+        self._filter(*args, **kwargs)
+        return self._clone()
+
+    def page_size(self, value):
+        if not value or not isinstance(value, six.integer_types):
+            raise SyncanoValueError('page_size value needs to be an int.')
+
+        self.query['page_size'] = value
+        return self._clone()
+
+    def limit(self, value):
+        if not value or not isinstance(value, six.integer_types):
+            raise SyncanoValueError('Limit value needs to be an int.')
+
+        self._limit = value
+        return self._clone()
+
+    def order_by(self, field):
+        if not field or not isinstance(field, six.string_types):
+            raise SyncanoValueError('Order by field needs to be a string.')
+
+        self.query['order_by'] = field
+        return self._clone()
+
+    def raw(self):
+        self._serialize = False
+        return self._clone()
+
+    # Other stuff
+
+    def contribute_to_class(self, model, name):
+        setattr(model, name, ManagerDescriptor(self))
+
+        self.model = model
+        if hasattr(model._meta, 'connection') and model._meta.connection:
+            self.connection = model._meta.connection
+
+        if not self.name:
+            self.name = name
+
+    def _filter(self, *args, **kwargs):
+        if args and self.endpoint:
+            properties = self.model._meta.get_endpoint_properties(self.endpoint)
+            mapped_args = {k: v for k, v in zip(properties, args)}
+            self.properties.update(mapped_args)
+        self.properties.update(kwargs)
+
+    def _clone(self):
+        # Maybe deepcopy ?
+        manager = self.__class__()
+        manager.name = self.name
+        manager.model = self.model
+        manager.connection = self.connection
+        manager.endpoint = self.endpoint
+        manager.properties = deepcopy(self.properties)
+        manager._limit = self._limit
+        manager.method = self.method
+        manager.query = deepcopy(self.query)
+        manager.data = deepcopy(self.data)
+        manager._serialize = self._serialize
+
+        return manager
+
+    def serialize(self, data):
+        if not isinstance(data, dict):
+            return
+        return self.model(**data) if self._serialize else data
+
+    def request(self, method=None, path=None, **request):
+        meta = self.model._meta
+        method = method or self.method
```

Comment:
I know that using state of the Manager, makes function calls shorten, but uh, I prefer more functional approach.